pkg/icingadb/runtime_updates.go

```diff
@@ -95,57 +95,176 @@ func (r *RuntimeUpdates) Sync(
+			entities := make(chan contracts.Entity, 1)
+			g.Go(func() error {
+				defer close(entities)
+
+				var counter com.Counter
+				defer periodic.Start(ctx, r.logger.Interval(), func(_ periodic.Tick) {
+					if count := counter.Reset(); count > 0 {
+						r.logger.Infof("Upserted %d %s sla lifecycles", count, s.Name())
+					}
+				}).Stop()
+
+				sl := v1.NewSlaLifecycle()
+				stmt, _ := r.db.BuildInsertIgnoreStmt(sl)
+
+				// Not to mess up the already existing FIFO mechanism, we have to perform only a single query
+				// (semaphore 1) at a time, even the sla queries could be bulk executed.
+				// After successfully upserting a lifecycle entity, the original checkable entity is streamed to "entities".
+				slaEntities := CreateSlaLifecyclesFromCheckables(ctx, g, r.db, s.Entity(), upsertEntities, false)
+				return r.db.NamedBulkExec(
+					ctx, stmt, upsertCount, semaphore.NewWeighted(1), slaEntities, com.NeverSplit[contracts.Entity],
+					OnSuccessApplyAndSendTo(entities, GetCheckableFromSlaLifecycle), OnSuccessIncrement[contracts.Entity](&counter),
+				)
+			})
 
-		g.Go(func() error {
-			var counter com.Counter
-			defer periodic.Start(ctx, r.logger.Interval(), func(_ periodic.Tick) {
-				if count := counter.Reset(); count > 0 {
-					r.logger.Infof("Deleted %d %s items", count, s.Name())
+			// Start the regular Icinga DB checkables upsert stream.
+			g.Go(upsertEntityFunc(entities))
+
+			updatedEntities := make(chan contracts.Entity, r.redis.Options.XReadCount)
+			g.Go(func() error {
+				defer close(updatedEntities)
+
+				var counter com.Counter
+				defer periodic.Start(ctx, r.logger.Interval(), func(_ periodic.Tick) {
+					if count := counter.Reset(); count > 0 {
+						r.logger.Infof("Updating %d %s sla lifecycles", count, s.Name())
+					}
+				}).Stop()
+
+				sl := v1.NewSlaLifecycle()
+				stmt := fmt.Sprintf(`UPDATE %s SET delete_time = :delete_time WHERE "id" = :id AND "delete_time" = 0`, utils.TableName(sl))
+				sem := r.db.GetSemaphoreForTable(utils.TableName(sl))
+
+				// Sometimes there are deletion events for checkables that do not have a "created at" sla lifecycle.
+				// Therefore, try to perform updates and stream the entities into "updatedEntities" where at least
+				// their delete event time will be tracked.
+				return r.db.NamedBulkExec(
+					ctx, stmt, upsertCount, sem, CreateSlaLifecyclesFromCheckables(ctx, g, r.db, s.Entity(), deleteEntities, true),
+					com.NeverSplit[contracts.Entity], OnSuccessSendTo(updatedEntities), OnSuccessIncrement[contracts.Entity](&counter),
+				)
+			})
+
+			deletedIds := make(chan interface{}, r.redis.Options.XReadCount)
+			g.Go(func() error {
+				// extractEntityId is used as a callback for the on success mechanism to extract the checkable ids.
+				extractEntityId := func(e contracts.Entity) interface{} {
+					sl, _ := e.(*v1.SlaLifecycle)
+					return sl.SourceEntity.ID()
```

Comment:
> This would panic if `e` isn't a `*v1.SlaLifecycle` with a nil dereference.

How should this ever happen? The `_` clearly indicates not to expect any errors, which implies that `e` being always `*v1.SlaLifecycle`. Though from a stylistic point of view, your suggestion looks way better.

Reply:
Yes, this should never happen (if I'd expect it to happen, I'd have mentioned that instead). But in this situation, there is really no point in using the non-panicking type assertion when the following would crash nonetheless if the assertion failed.

And if there was a bug, the panicking version would result in a more useful error, compare these two for exampe:
- https://go.dev/play/p/QQmKf-t5iPG
- https://go.dev/play/p/yB7My67Nlmz

One clearly says that there was something totally wrong in the interface variable while in the other case, it's unclear from the error alone whether this was due to the type assertion failing or if an actual nil pointer value somehow made it into the interface variable.